.github/workflows/nightly-release.yaml

```diff
@@ -0,0 +1,99 @@
+#     does the error go? Create a draft, unpublished release instead?
+#
+# - Build artifacts for all the different OSes we have GitHub runners for rather
+#   than just x86 Linux.
+
+name: Nightly Release
+
+on:
+  schedule:
+    - cron: '0 2 * * *'
+  # Enable manual runs for testing or manually (re-)creating a nightly release.
+  workflow_dispatch:
+
+permissions:
+  contents: write # For creating and uploading to releases.
+
+jobs:
+  release:
+    runs-on: ubuntu-22.04
+    steps:
+      - name: Harden Runner
+        uses: step-security/harden-runner@63c24ba6bd7ba022e95695ff85de572c04a18142 # v2.7.0
+        with:
+          egress-policy: audit
+
+      - name: Checkout branch
+        uses: actions/checkout@b4ffde65f46336ab88eb53be808477a3936bae11 # v4.1.1
+
+      - name: Set up remote cache access
+        env:
+          REMOTE_CACHE_KEY: ${{ secrets.CARBON_BUILDS_GITHUB }}
+        run: |
+          echo "$REMOTE_CACHE_KEY" | base64 -d > $HOME/remote_cache_key.json
+          echo "remote_cache_upload=--google_credentials=$HOME/remote_cache_key.json" \
+              >> $GITHUB_ENV
+
+      - uses: ./.github/actions/build-setup-ubuntu
+
+      - uses: ./.github/actions/build-setup-common
+        with:
+          matrix_runner: ubuntu-22.04
+          remote_cache_upload: ${{ env.remote_cache_upload }}
+
+      - name: Get nightly date
+        run: |
+          echo "nightly_date=$(date '+%Y.%m.%d')" >> $GITHUB_ENV
+
+      - name: Build release
+        env:
+          # 'libtool_check_unique failed to generate' workaround.
+          # https://github.com/bazelbuild/bazel/issues/14113#issuecomment-999794586
+          BAZEL_USE_CPP_ONLY_TOOLCHAIN: 1
+        run: |
+          ./scripts/run_bazel.py \
+            --attempts=5 --jobs-on-last-attempt=4 \
+            build -c opt --remote_download_toplevel \
+            --pre_release=nightly --nightly_date=${{ env.nightly_date }} \
+            //toolchain/install:prefix_root/bin/carbon \
+            //toolchain/install:carbon_toolchain_tar_rule
+
+      - name: Extract the release version
+        run: |
+          # Make sure we can run the toolchain to get the version.
+          ./bazel-bin/toolchain/install/prefix_root/bin/carbon version
+
+          # Now stash it in a variable and export it.
+          VERSION=$( \
+            ./bazel-bin/toolchain/install/prefix_root/bin/carbon version \
+            | cut -d' ' -f5 | cut -d'+' -f1)
+          echo "release_version=$VERSION" >> $GITHUB_ENV
+
+      - name: Create the release
+        env:
+          GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+        run: |
+          gh release create \
+            --title "Nightly build ${{ env.nightly_date }}" \
+            --notes 'A nightly development build of Carbon.' \
+            --prerelease \
+            v${{ env.release_version }} \
+            "bazel-bin/toolchain/install/carbon_toolchain-${{ env.release_version }}.tar.bz2#Linux x86-64 binaries - Ubuntu-22.04"
```

Comment:
What's your system for choosing these release asset names? Is there a project you're using as a baseline for the format?

To detail what I'm seeing, Carbon is: "Linux x86-64 binaries - Ubuntu-22.04"; actual file: carbon_toolchain-0.0.0-0.nightly.2024.06.17.tar.bz2

Comparing with some other projects that have release assets:

- https://github.com/llvm/llvm-project/releases has: clang+llvm-18.1.7-aarch64-linux-gnu.tar.xz
- https://github.com/bazelbuild/bazel/releases has: bazel-7.2.1rc1-linux-x86_64
- https://github.com/microsoft/TypeScript/releases/tag/v5.4.5 has: typescript-5.4.5.tgz

Concretely, my suggestion would be:

- Maybe drop "toolchain" since this is probably the normal install
- Add the target platform to the filename (this especially because it seems standard for platform-specific binaries)
- Consider removing the alias altogether (at least `Ubuntu-22.04` unless there's a reason for mentioning that)

i.e., make the asset name something like "carbon-${{ env.release_version }}-x86_64.tar.bz2"

Reply:
A few factors here...

Do we use a human-oriented description after the `#` or just the base filename? One benefit of a human-oriented name is that it gives us a good way to explain restrictions on where the binary artifacts will work that may not be apparent from a file name. Eventually, I'd like all of those to be embedded in the filename though, it's just not there yet today.

I think we do need to telegraph two restrictions on where these binaries can be used:
- Only on Linux machines, and not all Linux.
  - It needs to have glibc with at least the same version as that used on Ubuntu 22.04 (the host we build on).
- Only x86-64 machines.

The x86-64 restriction is one that I want to show up in the filename and I left a TODO about fixing this in the install code. But until then, I was trying to document it in the English alias. I suspect this is also the one where we'll start adding packages to cover the alternatives first.

The Linux and glibc version restriction is tricky. We're not building fully Linux-hermetic binaries at this point, and until we do, there is basically a "as-modern-as" requirement that is somewhat hard to document precisely in a way people understand. The clearest I came up with is "it was built on this OS version, so if a binary built for that OS works for you, you're good". Which is true, but maybe an odd way of explaining it. Happy to have better suggestions here for how we spell this. If it is something more than "linux", it may be tricky to wire through to the filename in Bazel, and so it may be something that has to be handled here in the action -- either renaming the file or adding an alias to clarify.

All this said, I'm happy to change this to whatever folks like for now. Just let me know, this isn't a case where I have any opinion or am worried about any of the alternatives really. They all seem fine.

Reply:
Looking through the top 10 C++ GitHub projects according to https://github.com/EvanLi/Github-Ranking/blob/master/Top100/CPP.md, half distribute assets; of those, all use filenames, and only gpt4all diverges in the naming scheme (swapping version and platform). For quick reference:

- https://github.com/electron/electron/releases/tag/v31.0.1
- https://github.com/microsoft/terminal/releases/tag/v1.20.11381.0
- https://github.com/godotengine/godot/releases/tag/4.2.2-stable
- https://github.com/opencv/opencv/releases/tag/4.10.0
- https://github.com/nomic-ai/gpt4all/releases/tag/v2.8.0
- https://github.com/protocolbuffers/protobuf/releases/tag/v27.1

So the filename naming scheme looks very common, and users should be expected be familiar with it. I expect it has not encountered readability issues, given the lack of label adoption. Maybe the lack of divergence indicates the filename structure has been working well?

Filenames do have other benefits beyond the simple consistency, such as knowing what filename to expect on download without further examination. If you still prefer a label, I'd suggest removing "binaries" or replacing it, because the library assets would probably remain part.

Note, for communicating the Ubuntu 22.04 requirement, you could add it as a suffix, such as "carbon_toolchain-0.0.0-0.nightly.2024.06.17.tar.bz2 (built on Ubuntu 22.04)". Just a thought for maintaining the filename idiom while still getting you the extra platform label.

Note though, I've approved. I don't think this is worth continuing to debate if you feel strongly.

Reply:
Nice data looking at the existing projects! =D

I really didn't have a strong opinion here, I'm happy to go with just the filename and work on getting Bazel to put more information there rather than stressing about adding it in annotations.